keps/sig-instrumentation/1668-trace-popagating/README.md

```diff
@@ -0,0 +1,501 @@
+
+```go
+// WithObject returns a context attached with a Span retrieved from object annotation, it doesn't start a new span
+func WithObject(ctx context.Context, obj meta.Object) (context.Context, error)
+```
+
+When controllers create/update/delete an object A based on another B, we propagate context from B to A. Below is an example to show how deployment propagate trace context  to replicaset.
+
+```diff
+ func (dc *DeploymentController) getNewReplicaSet(d *apps.Deployment, rsList, oldRSs []*apps.ReplicaSet, createIfNotExisted bool) (*apps.ReplicaSet, error) {
++       ctx := httptrace.WithObject(context.Background(), d)
+        existingNewRS := deploymentutil.FindNewReplicaSet(d, rsList)
+@@ -220,7 +227,8 @@ func (dc *DeploymentController) getNewReplicaSet(d *apps.Deployment, rsList, old
+        // hash collisions. If there is any other error, we need to report it in the status of
+        // the Deployment.
+        alreadyExists := false
+-       createdRS, err := dc.client.AppsV1().ReplicaSets(d.Namespace).Create(context.TODO(), &newRS, metav1.CreateOptions{})
++       createdRS, err := dc.client.AppsV1().ReplicaSets(d.Namespace).Create(ctx, &newRS, metav1.CreateOptions{})
+```
+
+#### Add Go context to parameter list
+In OpenTelemetry's Go implementation,  span context is passed down through Go context. This will necessitate the threading of context across more of the Kubernetes codebase, which is a [desired outcome regardless](https://github.com/kubernetes/kubernetes/issues/815). In alpha stage,  we need to change some APIs by adding `ctx context.Context` to parameter list whose parameters doesn't contain context.Context yet. Below APIs will be impacted so far.
+
+| APIs                          | file name                                                    |
+| ----------------------------- | ------------------------------------------------------------ |
+| createPods()                  | pkg/controller/controller_utils.go                           |
+| CreatePodsWithControllerRef() | pkg/controller/controller_utils.go<br />pkg/controller/replication/conversion.go<br />pkg/controller/daemon/daemon_controller.go<br />pkg/controller/replication/conversion.go |
+
+
+### Out-of-tree changes
+
+#### Mutating webhook
+We use mutating admission controller(aka webhook)  to change/update the object annotation. It takes advantages of:
+
+- Ease of use. Using client-go with a context.Context is easier than adding an annotation. The webhook takes care of writing the annotation.
+- Object to object context propagation. Without the mutating admission controller, we can only associate actions from a single object. With the mutating admission controller, the logging metadata would be added for objects modified by controllers of the initial object (e.g. metadata added to a deployment annotation would appear in pod logs).
```

Comment:
I'm not following this point. I would like to know what differentiates the webhook approach that enables us to fanout a trace across multiple objects.

Reply:
>> Object to object context propagation. Without the mutating admission controller, we can only associate actions from a single object. With the mutating admission controller, the logging metadata would be added for objects modified by controllers of the initial object (e.g. metadata added to a deployment annotation would appear in pod logs).

> I'm not following this point. I would like to know what differentiates the webhook approach that enables us to fanout a trace across multiple objects.

Indeed,  the 2nd point is confused here, it doesn't belong to an _advantage_ acutally.  What we want to express is that  "Object to object context propagation" is needed by storing the trace context to object via webhook. I'm trying to rewrite this part to make it more understandable. It would be appreciated if you could give some hints about this part.

Reply:
The mutating admission controller is required for _any_ object-based context propagation, even for a single object.  Without it, the context annotation is never written.

Object to object propagation is accomplished through the changes in controllers.  Using the `WithObject` function, you are extracting context from one object, and using it in the write for another object.

Reply:
Yes, this is why we need to use the mutating webhook, thanks for detailed explaination.